Unit test runner hides the normal pytest failure output

##### SUMMARY
When running `ansible-test units ...`, failed tests have no diagnostic messages attached to the assertion error.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ansible-test

##### ANSIBLE VERSION
```paste below
ansible 2.9.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/tadej/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/tadej/xlab/ansible/ansible_collections/debug/collection/venv/lib/python3.7/site-packages/ansible
  executable location = /home/tadej/xlab/ansible/ansible_collections/debug/collection/venv/bin/ansible
  python version = 3.7.6 (default, Jan 30 2020, 09:44:41) [GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]
```

##### CONFIGURATION
`ansible-config dump --only-changed` reported no changes.

##### OS / ENVIRONMENT
Tested on latest stable Fedora, Ansible was installed in a virtual environment using pip.

##### STEPS TO REPRODUCE
Run `ansible-test units path/to/test_file.py`, where *path/to/test_file.py* should have a failing test case akin to the next one:

    def test_x():
        assert dict(a=1) == dict(a=2)

See https://github.com/tadeboro/ansible-test-bad-assert for ready-to-run example.

##### EXPECTED RESULTS
In the case above, `pytest` prints this out:
```
========================== FAILURES ===========================
___________________________ test_x ____________________________

    def test_x():
>       assert dict(a=1) == dict(a=2)
E       AssertionError: assert {'a': 1} == {'a': 2}
E         Differing items:
E         {'a': 1} != {'a': 2}
E         Use -v to get the full diff

tests/unit/test_x.py:2: AssertionError
```


##### ACTUAL RESULTS
`ansible-test units` prints this:
```
========================== FAILURES ===========================
___________________________ test_x ____________________________
[gw1] linux -- Python 3.7.6 /tmp/python-hvsrfqdb-ansible/python
def test_x():
>       assert dict(a=1) == dict(a=2)
E       AssertionError

tests/unit/test_x.py:2: AssertionError
```

##### ADDITIONAL NOTES

I can bring back the diagnostic messages by commenting out the line https://github.com/ansible/ansible/blob/35996e57abba1f40abb493379d5cbe2dd90e65c3/test/lib/ansible_test/_data/pytest/plugins/ansible_pytest_collections.py#L40, but this of course then breaks other stuff.

--- a/changelogs/fragments/ansible-test-pytest-assertion-rewriting.yml
+++ b/changelogs/fragments/ansible-test-pytest-assertion-rewriting.yml
@@ -1,0 +1,5 @@
+bugfixes:
+  - ansible-test - Add support for ``pylint`` assertion rewriting when running unit tests on Python 3.5 and later.
+                   Resolves issue https://github.com/ansible/ansible/issues/68032
+known_issues:
+  - ansible-test - Unit tests for collections do not support ``pylint`` assertion rewriting on Python 2.7.

--- a/test/integration/targets/ansible-test-units-assertions/aliases
+++ b/test/integration/targets/ansible-test-units-assertions/aliases
@@ -1,0 +1,4 @@
+shippable/generic/group1  # runs in the default test container
+context/controller
+needs/target/collection
+needs/target/ansible-test

--- a/test/integration/targets/ansible-test-units-assertions/ansible_collections/ns/col/tests/unit/plugins/modules/test_assertion.py
+++ b/test/integration/targets/ansible-test-units-assertions/ansible_collections/ns/col/tests/unit/plugins/modules/test_assertion.py
@@ -1,0 +1,6 @@
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+def test_assertion():
+    assert dict(yes=True) == dict(no=False)

--- a/test/integration/targets/ansible-test-units-assertions/runme.sh
+++ b/test/integration/targets/ansible-test-units-assertions/runme.sh
@@ -1,0 +1,22 @@
+#!/usr/bin/env bash
+
+source ../collection/setup.sh
+
+set -x
+
+options=$("${TEST_DIR}"/../ansible-test/venv-pythons.py --only-versions)
+IFS=', ' read -r -a pythons <<< "${options}"
+
+for python in "${pythons[@]}"; do
+  if ansible-test units --color --truncate 0 --python "${python}" --requirements "${@}" 2>&1 | tee pylint.log; then
+    echo "Test did not fail as expected."
+    exit 1
+  fi
+
+  if [ "${python}" = "2.7" ]; then
+    grep "^E  *AssertionError$" pylint.log
+  else
+
+    grep "^E  *AssertionError: assert {'yes': True} == {'no': False}$" pylint.log
+  fi
+done

--- a/test/integration/targets/ansible-test/venv-pythons.py
+++ b/test/integration/targets/ansible-test/venv-pythons.py
@@ -1,6 +1,7 @@
 #!/usr/bin/env python
 """Return target Python options for use with ansible-test."""
 
+import argparse
 import os
 import shutil
 import subprocess
@@ -10,6 +11,11 @@ from ansible import release
 
 
 def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--only-versions', action='store_true')
+
+    options = parser.parse_args()
+
     ansible_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(release.__file__))))
     source_root = os.path.join(ansible_root, 'test', 'lib')
 
@@ -33,6 +39,10 @@ def main():
             print(f'{executable} - {"fail" if process.returncode else "pass"}', file=sys.stderr)
 
             if not process.returncode:
+                if options.only_versions:
+                    args.append(python_version)
+                    continue
+
                 args.extend(['--target-python', f'venv/{python_version}'])
 
     print(' '.join(args))

--- a/test/lib/ansible_test/_util/target/pytest/plugins/ansible_pytest_collections.py
+++ b/test/lib/ansible_test/_util/target/pytest/plugins/ansible_pytest_collections.py
@@ -32,6 +32,50 @@ def collection_pypkgpath(self):
     raise Exception('File "%s" not found in collection path "%s".' % (self.strpath, ANSIBLE_COLLECTIONS_PATH))
 
 
+def enable_assertion_rewriting_hook():  # type: () -> None
+    """
+    Enable pylint's AssertionRewritingHook on Python 3.x.
+    This is necessary because the Ansible collection loader intercepts imports before the pylint provided loader ever sees them.
+    """
+    import sys
+
+    if sys.version_info[0] == 2:
+        return  # Python 2.x is not supported
+
+    hook_name = '_pytest.assertion.rewrite.AssertionRewritingHook'
+    hooks = [hook for hook in sys.meta_path if hook.__class__.__module__ + '.' + hook.__class__.__qualname__ == hook_name]
+
+    if len(hooks) != 1:
+        raise Exception('Found {} instance(s) of "{}" in sys.meta_path.'.format(len(hooks), hook_name))
+
+    assertion_rewriting_hook = hooks[0]
+
+    # This is based on `_AnsibleCollectionPkgLoaderBase.exec_module` from `ansible/utils/collection_loader/_collection_finder.py`.
+    def exec_module(self, module):
+        # short-circuit redirect; avoid reinitializing existing modules
+        if self._redirect_module:  # pylint: disable=protected-access
+            return
+
+        # execute the module's code in its namespace
+        code_obj = self.get_code(self._fullname)  # pylint: disable=protected-access
+
+        if code_obj is not None:  # things like NS packages that can't have code on disk will return None
+            # This logic is loosely based on `AssertionRewritingHook._should_rewrite` from pytest.
+            # See: https://github.com/pytest-dev/pytest/blob/779a87aada33af444f14841a04344016a087669e/src/_pytest/assertion/rewrite.py#L209
+            should_rewrite = self._package_to_load == 'conftest' or self._package_to_load.startswith('test_')  # pylint: disable=protected-access
+
+            if should_rewrite:
+                # noinspection PyUnresolvedReferences
+                assertion_rewriting_hook.exec_module(module)
+            else:
+                exec(code_obj, module.__dict__)  # pylint: disable=exec-used
+
+    # noinspection PyProtectedMember
+    from ansible.utils.collection_loader._collection_finder import _AnsibleCollectionPkgLoaderBase
+
+    _AnsibleCollectionPkgLoaderBase.exec_module = exec_module
+
+
 def pytest_configure():
     """Configure this pytest plugin."""
     try:
@@ -39,6 +83,8 @@ def pytest_configure():
             return
     except AttributeError:
         pytest_configure.executed = True
+
+    enable_assertion_rewriting_hook()
 
     # noinspection PyProtectedMember
     from ansible.utils.collection_loader._collection_finder import _AnsibleCollectionFinder

